enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,360 @@
+    WorkAgent -->> WorkAgent: Decode spec event and apply manifestwork
+```
+
+Event schema:
+
+```json
+{
+    "specversion": "1.0",
+    "id": "<cloud-events-message-id>",
+    "source": "<source-id>",
+    "type": "io.open-cluster-management.v1.manifestwork.specresync.requested”",
+    "datacontenttype": "application/json",
+    "time": "2018-04-05T17:31:00Z",
+    "data": {
+       resourceVersions: [
+          {
+           "resourceID": "748d04ad-887a-43aa-b123-da1373227011",
+           "resourceVersion": "3"
+          },
+          {
+            "resourceID": "858d03ad-997b-52bc-b456-ab1473228022",
+            "resourceVersion": "1"
+          }
+        ]
+    }
+}
+```
+
+### Status Resync Event
+
+Event Scenarios:
+
+1. When a source restarts, it sends a status resync event to the broker to synchronize all resource status and prevent data loss. If the component has a database, it calculates the status hash of all resources stored in the database and includes it in the event. If the component does not have a database, it sends an status resync event with empty payload.
+2. When a source reconnects to the broker after a disconnection, it sends a status resync event including the status hash of all its owned resources to ensure synchronization and prevent data loss.
+3. (Optional) A source may periodically send status resync events to mitigate unexpected errors.
+
+Event transfer flow:
+
+1. The work agent subscribes to the status resync event topic `/sources/+/resync/clusters/manifestsstatus`, for example: `/sources/+/resync/clusters/manifestsstatus`.
+2. The source calculates the hash of all resources it owns to generate a status hash.
+3. The source sends a status resync event with the resources hash as the event payload to the broker.
+4. Upon receiving the status resync event, the work agent responds by sending resource status events to the broker as follows:
+  - If the event payload is empty, the work agent returns the status of all resources it maintains.
+  - If the event payload is not empty, the work agent retrieves the resource with the specified ID and compares the received resource status hash with the current resource status hash. If they are not equal, the work agent resends the resource status message.
+
+```mermaid
+sequenceDiagram
+    WorkAgent -->> Broker: Subscribe topic - /sources/+/resync/clusters/manifestsstatus
+    Source -->> Broker: Subscribe topic - /sources/<source-id>/clusters/+/manifestsstatus
+    Source -->> Source: Generate status hash map and encode as status resync event
+    Source -->> Broker: Publish status resync event
+    Broker -->> WorkAgent: Send status resync event
+    WorkAgent -->> WorkAgent: Compare the status hashes in event message and agent and generate status event
+    WorkAgent -->> Broker: Publish status event for manifestwork
+    Broker -->> Source: Send status event for manifestwork
+    Source -->> Source: Decode status event and retrieve manifestwork status
+```
+
```

Comment:
should add a section on failure handling.

Also what is the impact of message resync?

Reply:
Added failure section(mainly about resync) in https://github.com/open-cluster-management-io/enhancements/pull/97/commits/dccf37032d8d9df6b9434ac1e7761a0dd37405c4

Reply:
Where and how is source going to store the result of `Decode status event and retrieve manifestwork status` . We need to make that clear.

Reply:
It is source `side` responsibility to maintain shadow status of the agent. So if source specifically is disconnected from the broker, there should be other components on the source `side` which should maintain that state. I think we should try to avoid pushing the burden of this onto the agent. This is the role that https://github.com/kube-orchestra/maestro/ plays. WDYT

Reply:
I think maestro may need to send a status resync message to agent if restart or under certain circumstance. status resync might not be required, but I think it should be a supported message and agent should respond to the message properly. It is broker or server's decision to send or not send the status resync message.

Reply:
i am with you. Thanks